Add cuda_ipc channel to TensorPipe

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46791 Add cuda_ipc channel to TensorPipe**
* #46827 Let JIT unpickler to accept CUDA DataPtr from read_record_

Differential Revision: [D25237121](https://our.internmc.facebook.com/intern/diff/D25237121)

mrshenli committed:
Add cuda_ipc channel to TensorPipe

[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli commented:
@lw this does not look great to me, but it seems like `C10_DECLARE_REGISTRY` does not like template types. Another option is to use union in `ChannelRegistration`. Let me know if you prefer that.

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -11,12 +11,21 @@
 #include <c10d/Store.hpp>
 #include <torch/csrc/distributed/rpc/rpc_agent.h>
 
+#include <tensorpipe/tensorpipe.h>
```
I think that for now we're trying to keep TensorPipe as a private dependency, which means we can't include its headers in PyTorch's private headers. I also don't think this is really needed, as we seem to only need it for `HAD_CUDA_IPC_CHANNEL`, and we can probably do without that.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -11,12 +11,21 @@
 
 namespace tensorpipe {
 
+#define TENSORPIPE_USE_CUDA_IPC_CHANNEL TENSORPIPE_HAS_CUDA_IPC_CHANNEL && defined(USE_CUDA)
```
I believe TENSORPIPE_HAS_CUDA_IPC_CHANNEL will always be false when TENSORPIPE_USE_CUDA is false, which in turn should have the exact same value as PyTorch's USE_CUDA. Which means that this macro could be dropped and replaced by just TP_HAS_CUDA_IPC_CHANNEL.

@beauby Can you confirm?

mrshenli replied:
Can `TENSORPIPE_HAS_CUDA_IPC_CHANNEL` be false when `USE_CUDA` is true?

lw replied:
No, sorry, I meant it the other way around. Mathematically, `TENSORPIPE_HAS_CUDA_IPC_CHANNEL` implies `USE_CUDA` (because the CUDA IPC channel requires CUDA and couldn't exist without it). The other direction however doesn't hold: you could have `USE_CUDA` _without_ having `TENSORPIPE_HAS_CUDA_IPC_CHANNEL` (for example on Windows, I believe, because IPC handles are Linux-only).

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -11,12 +11,21 @@
+
 class CpuBuffer;
+
+#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
+class CudaBuffer;
+#endif
```
Once TensorPipe will use multiple CUDA channels (which is actually already the case, there's cuda_xth too now) we'll need CudaBuffer to be forward-declared if any of them is present. Or, in fact, more generally, we should declare it whenever TENSORPIPE_USE_CUDA is enabled, and that flag should be equal to PyTorch's USE_CUDA.
```suggestion
#if USE_CUDA
class CudaBuffer;
#endif
```

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -34,6 +43,11 @@ namespace channel {
 class Context;
 using CpuContext = Context<CpuBuffer>;
+
+#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
+using CudaContext = Context<CudaBuffer>;
+#endif
```
Same here: we should make this depend on the general availability of CUDA, not of the IPC channel.
```suggestion
#if USE_CUDA
using CudaContext = Context<CudaBuffer>;
#endif
```

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -163,11 +172,26 @@ std::unique_ptr<ChannelRegistration> makeMultiplexedUvChannel() {
     mpt_uv,
     makeMultiplexedUvChannel);
 
+#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
```
```suggestion
#if TENSORPIPE_HAS_CUDA_IPC_CHANNEL
```

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -297,15 +322,26 @@ void TensorPipeAgent::startImpl() {
         priority, std::move(key), std::move(reg->channel));
   }
 
+#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
```
```suggestion
#if USE_CUDA
```

mrshenli replied:
this means `USE_CUDA` and `TENSORPIPE_HAS_CUDA_IPC_CHANNEL` must always set to the same value?

lw replied:
The point of this comment was that the code below must run if there is _any_ CUDA channel, even if it is not exactly the IPC one. Don't you think?

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -297,15 +322,26 @@ void TensorPipeAgent::startImpl() {
+#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
+
+  for (auto& key : TensorPipeCudaChannelRegistry()->Keys()) {
+    std::unique_ptr<CudaChannelRegistration> reg =
```
The CPU version of this code has some logic to allow users to override the priorities of channels through a (hidden) field of the options struct. Should we add the same logic here?

lw reviewed (approved):
LGTM! And sorry it took me so long to get to this, I must have missed it.

The only remark I have is that we should probably make sure to use the right macros to guard the parts that are specific just to the IPC channel vs. any CUDA channel.

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"




[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"


Differential Revision: [D25237121](https://our.internmc.facebook.com/intern/diff/D25237121)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -309,36 +311,52 @@ void TensorPipeAgent::startImpl() {
+        context_->registerChannel(
+            priority, std::move(key), std::move(reg->channel));
+      }
+  );
```
Nit: If you want to merge these two codepaths even further, there's a trick to do so. In principle we're dealing with different types so we would need templates, but templated lambdas are only available in C++-20 (the syntax is `[captures] <typename T> (params) { ... }`). However C++-14 has something that is almost the same, called generic lambdas, which are those that use `auto` for one of their params.

Hence you could change the `registerChannel` lambda to accept a `auto& registry` param, so that you can call it twice, once with `TensorPipeCpuChannelRegistry()` and once with `TensorPipeCudaChannelRegistry()`. (Inside that lambda, then, you'd have to replace `std::unique_ptr<FooChannelRegistration>` with `auto`, which is a bit less readable, but maybe we can add a comment).

mrshenli replied:
great, thanks, let me try this!

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -79,7 +76,7 @@ struct CpuChannelRegistration {
 C10_DECLARE_REGISTRY(TensorPipeCpuChannelRegistry, CpuChannelRegistration);
 
 struct CudaChannelRegistration {
```
Would it be possible to define a template `ChannelRegistration<TBuffer>` only once and then instantiate it twice with the two buffer types? I remember you said that registries don't like templates, is it this what you were referring to?

mrshenli replied:
This will run into the same issue of introducing more ifdefs as `ChannelRegistration<CudaContext>` won't exist if CUDA is not available. And I recall I tried this in the first attempt, but `C10_DECLARE_REGISTRY` doesn't seem to like it. Didn't spend too much time to dig.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -79,7 +76,7 @@ struct CpuChannelRegistration {
 
 struct CudaChannelRegistration {
-#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
+#ifdef USE_CUDA_NOT_ROCM
```
I'm wondering if there's a reason you're guarding just the _content_ of this struct with the #ifdef, rather than guarding the definition of the whole struct and the declaration of the corresponding registry. In other words, do we need the registry declaration to exist when CUDA is disabled?

mrshenli replied:
if we guard the entire struct, we will need more `ifdef` to guard `C10_DECLARE_REGISTRY`, `C10_DEFINE_REGISTRY`, and where we check channels. I am OK either way, let me know which way you prefer. I can add a followup PR to update that.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -178,7 +178,9 @@ C10_REGISTER_CREATOR(
-#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
+#if TENSORPIPE_HAS_CUDA_IPC_CHANNEL && defined(USE_CUDA_NOT_ROCM)
+
+constexpr int64_t kCudaIpcChannelPriority = 300;
```
Nit: I agree that this constant is only needed when the CUDA IPC channel is enabled, and thus it could be guarded inside the `#if`, but I still think it would be nice to have all the priorities grouped together since it's useful to have an overview and compare them.

lw reviewed (approved):
I've left some minor comments, feel free to ignore.

Also, as you've now replaced USE_CUDA with USE_CUDA_NOT_ROCM, I'm wondering if this should be updated too:
https://github.com/pytorch/pytorch/blob/ccd20e995fb56a1aafefd90c0f2c68c537534c67/cmake/Dependencies.cmake#L1328

lw commented on caffe2/CMakeLists.txt:
```diff
@@ -310,6 +310,9 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       add_dependencies(tensorpipe_agent torch c10d)
       if(USE_TENSORPIPE)
+        if(USE_CUDA AND (NOT USE_ROCM))
+          target_compile_definitions(tensorpipe_agent PUBLIC USE_CUDA_NOT_ROCM)
```
Actually, wait a second, USE_CUDA_NOT_ROCM is a macro that is only defined here, nowhere else? That might be a problem, since the CMake target `tensorpipe_agent` is _only_ used to build the C++ tests. The "real" PyTorch libraries don't use the `tensorpipe_agent` target. So I'm worried that when building the real PyTorch the macro USE_CUDA_NOT_ROCM won't be set and thus all the CUDA-specific code will be disabled.

mrshenli replied:
> That might be a problem, since the CMake target `tensorpipe_agent` is only used to build the C++ tests. The "real" PyTorch libraries don't use the `tensorpipe_agent` target.

Thanks for the catch. This is new to me. Then let me try `__HIP_PLATFORM_HCC__` instead.

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"


Differential Revision: [D25237121](https://our.internmc.facebook.com/intern/diff/D25237121)

[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"


Differential Revision: [D25237121](https://our.internmc.facebook.com/intern/diff/D25237121)

[ghstack-poisoned]

mrshenli committed:
Update on "Add cuda_ipc channel to TensorPipe"


Differential Revision: [D25237121](https://our.internmc.facebook.com/intern/diff/D25237121)

[ghstack-poisoned]
